### ISSUE TITLE
writeMarkdown does not include metadata - writeMarkdown(readMarkdown a) = a ?

### ISSUE THREAD

**Issue by andrewufrank** [jgm/pandoc#6402]
It is highly desirable that `write . read = id` and readMarkdown and writeMarkdown with suitable options should (and could) do this. It would be possible to build small tools manipulating the text in a markdown file without altering the metadata (e.g. automatic some spell check corrections, e.g. US to british spelling).

I have tried but cannot find the right combination of options and extension, to have the correctly read  metadata put out with the writeMarkdown. Thus, I am not certain if this is a bug report or a feature request.

The reader and writer  options are 
```

let readOptions = def{readerStandalone = True  
                                        ,  readerExtensions = extensionsFromList 
                                                [Ext_yaml_metadata_block ] 
                                        }   

let writeOptions =  def {writerSetextHeaders = False
                                        , writerExtensions = extensionsFromList 
                                                [Ext_yaml_metadata_block]                                            
                                        }
```
Here a minimal small working example, done with lts-15.13 and pandoc 2.9.1.1. Any help appriciated to find a solution!

[pandocTest.cabal.txt](https://github.com/jgm/pandoc/files/4683991/pandocTest.cabal.txt)
The file is a `.hs` file, just rename!
[PandocTest.txt](https://github.com/jgm/pandoc/files/4683963/PandocTest.txt)

**Comment by mb21** [comment]
If you're missing metadata, you most likely forgot the `--standalone` option?

Generally, yes round-tripping markdown is something we try to achieve... if that was the question?

Also, for questions, please use the [pandoc-discuss mailing list](https://groups.google.com/d/forum/pandoc-discuss)...

**Comment by tarleb** [comment]
Closing in favor of the corresponding [StackOverflow question](https://stackoverflow.com/q/62027734).

**Comment by tarleb** [comment]
On second thought, maybe we should open a new issue for this. The question *does* hint at usability problems with the API.

**Comment by andrewufrank** [comment]
I had some help from users of stackoverflow and got it working. There were two issues:

1.  it worked only (in the sense of including metadate in the output) when I used a template, compiled and  added it to the writerOptions:

```
let writeOptions =  P.def {P.writerSetextHeaders = False
                                , P.writerExtensions = P.extensionsFromList 
                                        [P.Ext_yaml_metadata_block] 
                                ,  P.writerTemplate = Just t3                                                 
                                }
```
It would be nice if just `roundtrip` could be achieved with the default template and the default writer.

2. I copied the template from [](https://github.com/jgm/pandoc-templates/blob/master/default.markdown) and found it added a space in front of the metadata block. The reason was
the space in the first line 
`$if(titleblock)$ $titleblock$` which should be:
`$if(titleblock)$$titleblock$`
then it worked. 

I guess the issue is one of documentation or a good example. I will put my solution on stackoverflow and hope it helps. Thank you all for a great tool!

**Comment by jgm** [comment]
If writerTemplate is Nothing, you will get a fragment. Since metadata is added via the template, that means no metadata.  Hope that's clear enough.

As for your point 2, I don't understand it. There is no space there.
There's a line break immediately after `$if(titleblock)$` in the default template.  Something must have happened in your copy/paste operation.

**Comment by andrewufrank** [comment]
It was suggested I could use a filter - which would reduce the need for clarify the use of a `round trip` functionality. When I checked this more, I found that my use case -- think of an application similar to a spell checker for known misspellings -- requires to read in a file against which each word is checked. I cannot see how a filter could include IO. This gives a clear separation when filters are to be used and when a read - process - write roundtrip function is needed.
